### ISSUE TITLE
Send User-Agent as part of request from OpenLocate

### ISSUE THREAD

**Issue by kurabi** [OpenLocate/openlocate-android#48]
The `User-Agent` should contain the following:

- Host App Name
- Host App Version
- Host App Package Name
- Host App Build Number
- OS Version
- SDK Name
- SDK Version

For example, on iOS, this is what ends up being sent:
`iOS Example/1.0 (org.openlocate.iOS-Example; build:1; iOS 11.0.0) OpenLocate/1.0.0`

**Comment by jaakla** [comment]
OS generates normally own User-Agent in HTTP lib level, and that content is also informational and useful. I would not suggest to try to override/overwrite it as it could be used with DeviceAtlas and others for lookup. If you need to send these extra data (OpenLocate SDK version is useful indeed), then why not to add them to the ordinary JSON data body?

**Comment by jaakla** [comment]
I checked what the User-Agent has now, automatically. for Android I have following: `Dalvik/2.1.0 (Linux; U; Android 8.1.0; Nexus 5X Build/OPM4.171019.016.A1)`. This has mostly the info you're looking for, except Application ID and SDK version. 

I would suggest to avoid replacing already provided User-Agent header with app/SDK-defined value, as these existing values are something you cannot get easily otherwise.  In fact you cannot get all the same data in app level as far as I know. The format is not convenient to be read, different vendors have different User-Agent formats and contents, but there are lookup solutions (like DeviceAtlas) to resolve this issue; and if you overwrite User-Agent then in fact you cannot use DeviceAtlas-type databases at all. Of course it is useful to get the additional data about SDK, device and app, but better use another custom header or headers (one per value), if you really do not want to put it to the body JSON.

p.s. I see that for iOS you already use custom header, in my sample `iOS Example/1.0.0 (com.openlocate.ios.example-app-cto; build:2018.04.04.23.09; iOS 11.3.1) OpenLocate/1.3.1`. I would suggest here same approach - keep standard header value, which would be more cryptic, but with extra info: `User-Agent: iOS%20Example/2018.04.04.23.09 CFNetwork/897.15 Darwin/17.5.0` in my quick test, and use another header for the custom "user agent" data.

**Comment by kurabi** [comment]
As you confirmed, the user agent strings are very inconsistent in android devices, they vary by manufacturer, android os version, and also between different carrier version, network stacks, etc. So we aren’t just adding a few fields, it is about also ensuring consistency across all various devices so that data doesn’t need to have a dependency on external non-open source proprietary provider. If you think we should add more fields then let us know. Otherwise you can include the default user agent in a custom header field.

**Comment by simonatsafegraph** [comment]
@jaakla -- would love to hear what fields you think the OS provides that could be valuable! I think the concern is that the user-agent available by default is so inconsistent, while the one we have constructed meets the specification outlined here: https://tools.ietf.org/html/rfc7231#section-5.5.3.

Device Atlas is a paid service so I think there are also concerns there about making the standardisation of the user-agent field dependent upon that.

Let us know what you think.

**Comment by jaakla** [comment]
DeviceAtlas here is just one sample of database of user-agents, [there are also others](https://stackoverflow.com/questions/1457380/is-there-an-online-user-agent-database). Yes, the values are hard to parse, probably providers do not follow the RFC, or the RFC itself is not strict enough. This is one reason why UserAgent database is needed. But this database is needed not only for basic data lookup, they also provide a lot of secondary data about same devices, e.g. screen size, cpu/gpu properties etc which are not in UserAgent. There can be even tertiary data like device popularity and pricing in certain markets, so you can indirectly get some idea about user profile. The UserAgent string is not really parsed, this is used as unique key for device model. One more advantage of the original UserAgent value is that it is same across different apps, which is very useful for larger data correlations. Here are limitations also - the user agent of browsers and HTTP lib is different, and lookup databases mostly focus on web browsers, so our success here may vary; and even more - iOS _CFNetwork_ provided useragent is already customized for app, so in practise Apple devices give too limited data there.

AI totally agree it makes sense to add same data as well-defined form, not expecting clients to be able to parse values from device-provided one, or making mandatory to use whatever lookup database or service. The question is just where to put it. From data receivers perspective the best form would be additional fields in existing JSON body, so I have just one parser for everything, not separate one for the headers. Also at least theoretically this would some bits, as body can be gzipped in http level using `Content-Encoding: gzip` (once you implement it), unlike headers. But if you want to put it to custom headers, then server can live with it also, just few extra lines of code to read it from there. In the past I've seen network proxies/gateways/firewalls of mobile operators who parse or even remove "unknown" headers, to laundry traffic, so I would feel body to be safer in this sense, but I believe nowadays where most use https (especially with such data collection) and this is unlikely.

> Otherwise you can include the default user agent in a custom header field.
Can you give some code samples? Last time I checked I could not find any device API on Android or iOS to get the original user agent value. This is something what HTTP stack adds automatically to requests. I imagine some indirect hacks using web server to get it, but better keep things simple. This is why I'd like to keep the value of it.

**Comment by jaakla** [comment]
p.s. the json body already several device/user fields, it is not just location data, like these:

		{
			"ad_id": "f109e57e-a02e-41d3-b7c4-c906d1b92331",
			"ad_opt_out": false,
			"id_type": "aaid",
			"device_manufacturer": "motorola",
			"device_model": "Moto G (5S) Plus",
			"os_version": "Android 7.1.1",
			"carrier_name": "T Mobile",
		}
